### ISSUE TITLE
How to make the libigl repository more lightweight?

### ISSUE THREAD

**Issue by jdumas** [libigl/libigl#804]
Hi there,

I'd like to open a topic to discuss how to make the libigl repository more lightweight.
Right now a fresh clone of the main libigl repository is around ~900 MB. We have:
- 600 MB in the `.git/` folder (not much we can do about that).
- 144 MB in `external/`
- 55 MB of assets in `tutorial/shared/`
- 47 MB of images in `docs/images/`
- 7.5 MB of code in `include/igl/` (I believe this is only text)

This is problematic when libigl is used as a submodule in a lot of other projects. You automatically need an extra 1 GB per project and that's a lot of disk space in the end.

While there is not much we can do about the history (due to git's conservative nature), we should try to alleviate the burden for future development and other projects using libigl as a dependency (e.g. shallow submodules can be used to not copy the whole history). It's be nice to discuss in this thread how to achieve this. Here are my thoughts below:

- For `external/` we have 96 MB coming from `CoMiSo`. Most of it seems to be precompiled dll for Windows. Maybe we can drop the submodule (and only keep the most useful ones by default). We can also use CMake to download the precompiled dlls on Windows, so that the files don't have to be bloat up the git repository and still provide a seamless experience on Windows. I'd be happy to look into it. Then we got another 10 MB from `AntTweakBar` that we could probably live without.

- The assets in the `tutorial/shared` folder could be kept in an external submodule, or downloaded by CMake at compilation time (e.g. by cloning another git repo, wihtout said repo being an explicit submodule).

- The `docs/images` folder is getting pretty heavy too, and updating or adding new tutorials/screenshots is not gonna help that. Now that we have a pretty website and the doc is completely isolated, maybe we could keep it in a separate repository (we could use an [Organization Pages site](https://help.github.com/articles/user-organization-and-project-pages/) with a repo `libigl.github.io` for example). This mean the website would be accessible directly under `http://libigl.github.io/`, as opposed to `http://libigl.github.io/libigl/` presently. Keeping the doc in a separate repo has other advantages e.g. you can give write access to more people for them to contribute.

**Comment by alecjacobson** [comment]
I'm _very much_ in favour of this goal.

A lot can and should be done. I think we may have to give up on some of the `.git/` folder to make any worthwhile dent.

Then we should have strict guidelines for assets/images (if they remain in the main repo).

I'm experimenting with bfg and other ways to strip down the library at https://github.com/alecjacobson/libigl-skinny

One thing I'd like to keep is the edit/contribution history. Although we might just need to bite the bullet and move this to a museum repo.

**Comment by jdumas** [comment]
Ah yes, I was wondering what was this libigl-skinny :p 

The danger with BFG and other tools that rewrite history (sounds like a very scary thing), is that it will invalidate the SHA1 number of **every** commit in the history. Thus if you replace the current libigl repository with this, every project that ever used libigl as a submodule will become invalid.

The only way to stay *backward-compatible* would be to move things to a new repo (`libigl2` or with a new name), in order not to break existing projects.

I still believe we should move assets and images to an external repo, as updating the assets will (e.g. if we redo the screenshots without the GUI) will only increase the history and the problem might pop up again in the future. Having the doc in a different repo has other advantages, and it's very easy to download the assets when compiling the tutorials.

**Comment by alecjacobson** [comment]
I'm ok with not staying backward-compatible in this way. We need to do something _major_. The status quo is not sustainable.

Obviously if we break backward compatibility. We should try to minimize damages and do things "right from now on."

What are the possible pros and cons of each of the proposed additional changes:

### moving tutorial/shared assets to separate (non-submoduled) repo

#### pro

- save 20 mb for anyone who does not want to use tutorial (e.g., non-first-time users)

#### cons

- building the tutorial will require an internet connection. This is clearly frustrating if you didn't realize that when you downloaded libigl. 
- building the tutorial will require git (perhaps unlikely that anyone is using libigl without git, but something to consider)

#### worry

- how stable / universally supported is pulling from a git repo via cmake? Will this work for the average windows user? Does it require fancy/non-standard cmake setup?

#### Current Conclusion

I'm not sure the benefit out ways the cost. If we can really strip down the assets to <20MB then I'd rather hunt for other savings. I'd even rather change the models we use in the tutorial to increase redundancy/savings. Keeping the tutorial near-zero-barrier-to-entry was a big effort (mostly by @danielepanozzo ) and I'd like to maintain that.

### moving docs to separate (non-submoduled) repo

#### pro

- save 20 mb for anyone who doesn't want the local docs (e.g., almost everyone?)

#### cons

- viewing the docs requires an internet connection or remembering to grab this extra repo

#### worry

- will we be able to continue using the url http://libigl.github.io/libigl/

#### Current Conclusion

This seems like a good option as long as the url http://libigl.github.io/libigl/ doesn't change.

### Hosting our own repos for submodules in `external/`  (e.g., instead of https://github.com/eigenteam/eigen-git-mirror we'd use https://github.com/libigl/eigen)

#### pro

- potentially **_big_** savings. Assuming we can strip each module's .git folder to bare essentials, I'm guessing saving something like 150MB
- we still get to ensure that all --recursive clones of libigl get the "officially supported" commit of each submodule

#### cons

- every time we want to update an external library we will have to do more work, we can't simply pull and commit a new link
- in our mirrored repos we need to make sure the .git folder doesn't just grow again over time: we have to be paranoid about keeping those clean when merging with new updates

#### Current conclusion

I think we should do this for the major offenders: Eigen, Embree and maybe imgui. Changing Eigen versions will _always_ be something requires extreme care. Embree is a sort of one-off and last time they had a major change we needed to rewrite our wrappers anyway. I don't know enough about imgui to know why its repo is so bloated or if it is stable enough to do this.

**Comment by alecjacobson** [comment]
^--- if CGAL has really fixed their cmake situation we might be soon tempted to include it as a submodule as well and this last conversation would have to apply to their repo, too

**Comment by jdumas** [comment]
I think preserving compatibility is important. You want to be able to open a research project from 6 years ago and still be able to compile and run the code. If you cannot do that because we screwed up the SHA1, that'd be very unfortunate. That's why I was suggesting migrating to a new repo with a different name, with the new repo having the cleaned-up history of the old one.


### Tutorials

There is also the option of moving them to their own repository, which would use libigl as a submodule. Pro: everything is self-contained and building doesn't require git once the initial clone/download is done. Cons: increases the fragmentation and have to deal with PR in multiple repos when adding a tutorial.

Regarding your worry, downloading via CMake works pretty well, via [external projects](https://cmake.org/cmake/help/v3.0/module/ExternalProject.html). I made [this script](https://github.com/libigl/libigl/blob/master/external/CMakeLists.txt) to download and build CGAL & Boost on Windows, and have run it without any issue (it's now part of the AppVeyor build). I'm not even sure that CMake needs an implementation of git available. And you can even download files individually (see e.g. [this line](https://github.com/libigl/libigl/blob/master/external/download.cmake#L9) which downloads the dlls for gmp and mpfr to emulate what the CGAL installer is doing on Windows). The only downside to this would be to require an internet connection for the first build indeed.

### Website

It wouldn't be difficult to keep the current url valid. With the repo `libigl.github.io` the website would be directly under https://libigl.github.io, but we can either 1) put a redirect that preserves subfolders (if that's possible with html, since github doesn't allow `.htacces`?), or 2) put the pages in a `libigl/` subfolder so they appear under the https://libigl.github.io/libigl url.

### External libraries

- Given how often we update Eigen, I think it's fine to keep a barebone copy of the repository. In fact, we can probably just keep only the include folders `Eigen/` and `unsupported/` as well as the toplevel readme/licence files, etc. (and ditch the test/, doc/ etc.).

- For imgui, the main library is only a couple of files that we could keep directly in the libigl repo. Most of the weight comes from the `example/` and `misc/` folder that are not needed for users.

- Maybe for other "non-essential" repositories (embree, CoMiSO ...) we can rely on CMake external projects to enable the user to *optionally* download them. The script would just be doing a git clone and still be compiled on Travis, but it wouldn't simply be a default when you run `git clone --recursive` on the libigl repo.

**Comment by alecjacobson** [comment]
I would rather optimize for current and future users, rather than legacy projects. 

What will the error message look like and how they would have to fix it?

For example, what if we create a `libigl/libigl-museum/` _fork_ of libigl so the commit history is perfectly maintained. Could legacy projects be instructed to simply change the url of the repo when they see an error?

I do not like the idea of changing the main repo name or the url of the website. It'd be really great to avoid this.

The major con for the rest of the discussion seems to orbit around the ease of use for first time users to get up and running with the tutorial. I've enjoyed the current set up, but maybe external projects is not so crazy (after all our current set up requires a very good internet connection and still waiting some minutes to download everything; perhaps it's a reasonable trade to make it all faster, but decentralized a bit).

The point about PRs for new features and tutorials is worth considering. For tutorials there's an incentive to add the entry since it acts like an advertisement. I've had much less luck convincing people to add unit tests. Maybe because it's a separate repo, but maybe because it feels like extra work with little immediate reward.

**Comment by jdumas** [comment]
> For example, what if we create a libigl/libigl-museum/ fork of libigl so the commit history is perfectly maintained. Could legacy projects be instructed to simply change the url of the repo when they see an error?

Ah, that's actually a good suggestion. My concern was that if there is no existing alternative, then the user would need to figure what is the new SHA1 that he needs to point to with the new history ... and it's basically gonna be impossible to sort out.

I also expect further complications if we want to clean out the submodules histories (e.g. if we want to recreate a repo for CoMiSo). While it's probably fine to ask users to change old code's urls to point to `libigl-museum` instead of `libigl`, if the submodules are not recursively available anymore it's gonna be a disaster.

CMake external projects are convenient, but the only drawback that they have is that they cannot be used directly in a target-oriented build, because the download happens at compilation time (when you do `make` and not `cmake`). So it'd be perfect to download assets for the tutorials (since they are only needed to run the programs, not for compilation). For external libraries they are useful if you accept the idea of running an extra `cmake; make` step to perform the actual download (and on Windows you'd need to open a second Visual Studio project and compile it once). So for external dependencies I think `submodules` are still more convenient, but it would be acceptable imho for the modules which are not used very often (CoMiSo, embree ...) and for which an extra step would be acceptable. I don't know if there is a better solution, but this seems like an acceptable middle ground for me.


As for the unit tests well ... this is true what you say. I guess there's an advantage in keeping everything in the main repo, as it's easier to contribute one PR to one repo, as opposed to 3 PR to 3 different repos. But then we end up with humongous repos, and that makes it annoying to use as a submodule.

Text and code is pretty small, so ultimately I think the main library, tutorials and unit tests should be in one unique repo, without any asset (image, mesh or otherwise). The doc/website should be a separate repo. And assets for the tutorials and unit tests can be two different repository that are downloaded via CMake external projects.

**Comment by alecjacobson** [comment]
In the meantime, let's test out this libigl-museum idea on a fake repo to be sure it would really work.

**Comment by evouga** [comment]
Is there a reason y'all don't want to do tagged releases and provide .zips of those releases (perhaps both a "lite" and "full" version)?

1) I'd imagine most people do not want to track the bleeding-edge development changes, or submit pull requests, and for these people the .git is wasted space regardless of space;
2) As a downstream user of libigl I'd much rather tell people, "my code works with libigl version X.Y.Z; get the zip here:" than "git clone the latest version of libigl, check out commit 03536c4, then run these additional arcane commands to set up the right submodules and versions";
3) You can make it clear to your users which snapshots of the library you consider significant/stable.

**Comment by evouga** [comment]
> I think preserving compatibility is important. You want to be able to open a research project from 6 years ago and still be able to compile and run the code. 

This 100%. Nothing is more discouraging than finding some research code online, and realizing that it's probably easier to reimplement the whole thing yourself than figure out how to get it and its dependencies to compile.

**Comment by jdumas** [comment]
> Is there a reason y'all don't want to do tagged releases and provide .zips of those releases (perhaps both a "lite" and "full" version)?

I, for one, would very much like to see this happening. There is currently a file with a [release history](https://github.com/libigl/libigl/blob/master/docs/RELEASE_HISTORY.md) but it's not tied to any particular commit number or git tag so it's impossible to identify which is which.

**Comment by alecjacobson** [comment]
This is on the roadmap. It's probably something we'll do after thinning out
the repo size. For now, our recommendation is to use sub modules, so that
you can tell downstream users to git clone --recursive your code and it
will automatically grab the correct version of libigl.

-Alec

**Comment by jdumas** [comment]
I believe we can close this one now =)
